frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetail.tsx

```diff
@@ -141,9 +144,27 @@ export function ObjectDetailFn({
   className,
 }: ObjectDetailProps): JSX.Element | null {
   const [hasNotFoundError, setHasNotFoundError] = useState(false);
+  const [maybeLoading, setMaybeLoading] = useState(false);
   const prevZoomedRowId = usePrevious(zoomedRowID);
-  const prevData = usePrevious(data);
+  const prevData = usePrevious(passedData);
   const prevTableForeignKeys = usePrevious(tableForeignKeys);
+  const [data, setData] = useState<DatasetData>(passedData);
+
+  const pkIndex = useMemo(() => {
+    const pkCount = passedData?.cols?.filter(isPK)?.length;
+    if (pkCount !== 1) {
+      return undefined;
+    }
+    return passedData?.cols?.findIndex(isPK);
```

Comment:
Nit: it might be worth refactoring this block (154-158) into its own function `getPkIndexFromDatasetData` and writing a quick test for it